OpenCL_C.txt

```diff
@@ -6513,14 +6522,41 @@ This section specifies each general kind.
 --
 [source,c]
 ----------
-// Requires OpenCL C 2.0, or both the __opencl_c_atomic_order_seq_cst
+// Requires OpenCL C 3.0 or newer and both the __opencl_c_atomic_order_seq_cst
 // and __opencl_c_atomic_scope_device feature macros.
+void atomic_store(volatile __global A *object, C desired)
+void atomic_store(volatile __local A *object, C desired)
+
+// Requires OpenCL C 2.0, or all of the __opencl_c_generic_address_space,
```

Comment:
btw I am not sure this wording is correct but I think we use it everywhere across:

> Requires OpenCL C 2.0, or all of the __opencl_c_generic_address_space, __opencl_c_atomic_order_seq_cst and __opencl_c_atomic_scope_device feature macros.

I guess what we mean here is that we require OpenCL C v2.0 or the following features to be supported. However, in contrast to extensions we never gave name to features so there is no way to enumerate them unambiguously now. I think however this is just a wording issue that can be fixed later?

Reply:
> I guess what we mean here is that we require OpenCL C v2.0 or the following features to be supported i.e. require feature macro doesn't really explain precisely what is needed.

Yes, exactly.  The wording is deliberately terse because there are a lot of cases.

> I presume if feature macros are defined by the user then most this functionality will be available but this is not the case generally.

Now I'm not sure I understand your comment.  The implementation sets the feature macros, not the user.   I.e. this builtin will only be present for implementations that also set the listed feature macros.

> However, in contrast to extensions, we never gave names to features so there is no way to enumerate them unambiguously now. I think this is just a wording issue that can be fixed later?

There is a table at the start of the document, these macro names are all present there.

Reply:
There is also still this ticket about seeing if features can be made clearer (but without actually changing their meaning etc): https://github.com/KhronosGroup/OpenCL-Docs/issues/330

It's not done yet but it's in the 3.0 project.

Reply:
Yes I think it makes sense to cover this in https://github.com/KhronosGroup/OpenCL-Docs/issues/330.

So what I actually meant is that for extensions we used to define a name that was the same as extension macro normally (see OpenCL Extensions section 1.1).

Then we could write something like:
> The half scalar and vector types can only be used if the cl_khr_fp16 extension is supported andhas  been  enabled. 

`cl_khr_fp16` was referring to the name of the extension.

Currently we can not say "pipe can only be used if the __opencl_c_pipes feature is supported" because `__opencl_c_pipes` is a macro technically and not a feature. But describing that a type requires feature macro is not entirely correct or at least imprecise.

However, let's leave this for now and address this in #330.